rust/src/lockfile.rs

```diff
@@ -0,0 +1,291 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use c_utf8::CUtf8Buf;
+use serde_derive::{Deserialize, Serialize};
+use serde_json;
+use serde_yaml;
+use std::collections::HashMap;
+use std::path::Path;
+use std::{fs, io};
+use failure::Fallible;
+use failure::ResultExt;
+
+pub struct Lockfile {
+    #[allow(dead_code)] // Not used in tests
+    parsed: LockfileConfig,
+    serialized: CUtf8Buf
+}
+
+#[derive(PartialEq)]
+enum InputFormat {
+    YAML,
+    JSON,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct LockfileConfig {
+    packages: HashMap<String, Vec<(String, String)>>
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct PermissiveLockfileConfig {
+    #[serde(flatten)]
+    config: LockfileConfig,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+struct StrictLockfileConfig {
+    #[serde(flatten)]
+    config: LockfileConfig,
+}
+
+/// Parse a YAML/JSON lockfile definition.
+fn lockfile_parse_stream<R: io::Read>(
+    fmt: InputFormat,
+    input: &mut R,
+) -> Fallible<LockfileConfig> {
+    let lockfile: LockfileConfig = match fmt {
+        InputFormat::YAML => {
+            let lf: StrictLockfileConfig = serde_yaml::from_reader(input).map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("serde-yaml: {}", e.to_string()),
+                )
+            })?;
+            lf.config
+        }
+        InputFormat::JSON => {
+            let lf: PermissiveLockfileConfig = serde_json::from_reader(input).map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    format!("serde-json: {}", e.to_string()),
+                )
+            })?;
+            lf.config
+        }
+    };
+    Ok(lockfile)
+}
+
+/// Given a lockfile filename, parse it
+fn lockfile_parse<P: AsRef<Path>>(
+    filename: P,
+) -> Fallible<LockfileConfig> {
+    let filename = filename.as_ref();
+    let mut f = io::BufReader::new(open_file(filename)?);
+    let basename = filename
+        .file_name()
+        .map(|s| s.to_string_lossy())
+        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "Expected a filename"))?;
+    let fmt = if basename.ends_with(".yaml") || basename.ends_with(".yml") {
+        InputFormat::YAML
+    } else {
+        InputFormat::JSON
+    };
+    let lf = lockfile_parse_stream(fmt, &mut f).map_err(|e| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!("Parsing {}: {}", filename.to_string_lossy(), e.to_string()),
+        )
+    })?;
+    Ok(lf)
+}
+
+/// Open file and provide context containing filename on failures.
+fn open_file<P: AsRef<Path>>(filename: P) -> Fallible<fs::File> {
```

Comment:
Might be nice to have this live in a utility file, but not important.

Reply:
Not using rust anymore.